crates/oxc_linter/src/rules/unicorn/consistent_function_scoping.rs

```diff
@@ -0,0 +1,1233 @@
+            let mut parent_scope_ids = FxHashSet::default();
+            parent_scope_ids.insert(current_scope_id);
+            while let Some(parent_scope_id) = ctx.scopes().get_parent_id(current_scope_id) {
+                parent_scope_ids.insert(parent_scope_id);
+                current_scope_id = parent_scope_id;
+            }
+            parent_scope_ids
+        };
+
+        for reference_id in function_body_var_references {
+            let reference = ctx.symbols().get_reference(reference_id);
+            let Some(symbol_id) = reference.symbol_id() else { continue };
+            let scope_id = ctx.symbols().get_scope_id(symbol_id);
+            if parent_scope_ids.contains(&scope_id) && symbol_id != function_declaration_symbol_id {
+                return;
+            }
+        }
+
+        ctx.diagnostic(
+            OxcDiagnostic::warn("Function does not capture any variables from the outer scope.")
+                .with_help("Move this function to the outer scope.")
+                .with_label(reporter_span),
+        );
+    }
+}
+
+#[derive(Default)]
+struct ReferencesFinder {
+    is_parent_this_referenced: bool,
+    references: Vec<ReferenceId>,
+    in_function: usize,
+}
+
+impl<'a> Visit<'a> for ReferencesFinder {
+    fn visit_identifier_reference(&mut self, it: &oxc_ast::ast::IdentifierReference<'a>) {
+        self.references.push(it.reference_id().unwrap());
+    }
+
+    fn visit_this_expression(&mut self, _: &oxc_ast::ast::ThisExpression) {
+        if self.in_function == 0 {
+            self.is_parent_this_referenced = true;
+        }
+    }
+
+    fn enter_node(&mut self, kind: AstKind<'a>) {
+        if let AstKind::Function(_) = kind {
+            self.in_function += 1;
+        }
+    }
+    fn leave_node(&mut self, kind: AstKind<'a>) {
+        if let AstKind::Function(_) = kind {
+            self.in_function -= 1;
+        }
+    }
+}
+
+fn is_parent_scope_iife<'a>(node: &AstNode<'a>, ctx: &LintContext<'a>) -> bool {
+    if let Some(parent_node) = outermost_paren_parent(node, ctx) {
+        if let Some(parent_node) = outermost_paren_parent(parent_node, ctx) {
+            if matches!(
+                parent_node.kind(),
+                AstKind::Function(_) | AstKind::ArrowFunctionExpression(_)
+            ) {
+                if let Some(parent_node) = outermost_paren_parent(parent_node, ctx) {
+                    return matches!(parent_node.kind(), AstKind::CallExpression(_));
+                }
+            }
+        }
+    }
+
+    false
+}
+
+// https://reactjs.org/docs/hooks-reference.html
+const REACT_HOOK_NAMES: phf::Set<&'static str> = phf_set! {
+    "useState",
+    "useEffect",
+    "useContext",
+    "useReducer",
+    "useCallback",
+    "useMemo",
+    "useRef",
+    "useImperativeHandle",
+    "useLayoutEffect",
+    "useDebugValue"
+};
+
+fn is_in_react_hook<'a>(node: &AstNode<'a>, ctx: &LintContext<'a>) -> bool {
```

Comment:
Did you mean to use [`is_react_hook`](https://github.com/oxc-project/oxc/blob/4cb8c3771fa5448efd82a0c16413d62e334bd140/crates/oxc_linter/src/utils/react.rs#L277)?

Reply:
hmm i guess we could - it looks like that function reports on all react hooks whereas the unicorn rule only reports on the main builtin ones.

Reply:
This could work to our advantage, no?

Reply:
not sure, but i changed it anyway. we can always review based on feedback